lib/Dialect/TritonGPU/Transforms/Pipeliner/LowerLoops.cpp

```diff
@@ -1106,6 +1106,26 @@ scf::ForOp lowerMMA(ttng::MMAv5OpInterface mma, scf::ForOp forOp,
   }
 
   OpBuilder builder(forOp);
+
+  if (auto initFlagArg = dyn_cast<BlockArgument>(mma.useAccumulator())) {
+    // Undo the accumulator init optimization where the init flag is initialized
+    // with false and uncoditionally set to true after the first iteration,
+    // since this results in an incorrect IR after the MMA is peeled
+    auto idx = initFlagArg.getArgNumber() - 1;
+    auto yieldOp = mlir::cast<scf::YieldOp>(forOp.getBody()->getTerminator());
+    auto initArgs = forOp.getInitArgs();
+    auto yieldOperands = yieldOp.getOperands();
+    auto initUseAcc = getBoolFromConstant(initArgs[idx]);
+    auto yieldUseAcc = getBoolFromConstant(yieldOperands[idx]);
+
+    if (initUseAcc && *initUseAcc == false && yieldUseAcc &&
+        *yieldUseAcc == true) {
+      auto vTrue = builder.create<arith::ConstantOp>(forOp.getLoc(),
+                                                     builder.getBoolAttr(true));
+      mma.setUseAccumulator(vTrue);
+    }
+  }
```

Comment:
Without this, the pipeliner peels the first iteration MMA out of the loop, and leaves the MMA whose accum flag is initialized to false in the inner loop. A better way is to post-process the flag after things are expanded, but for now I settled on this simple workaround.

Reply:
I don't understand why the IR would be wrong due to peeling of the first iteration?

Reply:
Because the loop is left with the useAcc flag initialized to false. The MMA of the first iteration ends up clearing the result of the peeled MMA.

Reply:
isn't it what we want?

Reply:
Not sure what you mean, I'm saying that the following IR is incorrect

```
MMA acc // the peeled one 

for k ... iter_args(acinit_flag = false)
    MMA acc init_flag
    yield true
```

Without the workaround here, this is what we end up running and all blackwell matmul tests fail.

Reply:
ah I see what you mean, I had the flag meaning swapped. It is pretty concerning if the pipeliner generates wrong code out of such a simple case. Do you have the IR before and after?

Reply:
Before https://gist.github.com/masahi/2b6bbebf996658b6f30fa9c1c81562b4
After (generated without the WA) https://gist.github.com/masahi/f3796c6568b2a428a557fe24d51b11dc

The issue is clear. The pipeliner syntactically expands stuff, so it doesn't understand that the implication of peeling on the initial value of `init_flag`.

I was not sure if doing ad-hoc post-process of the init flag after expansion is a good idea, so I simply undo this optimization for SWP. In my use case, I don't go through SWP, so this is fine.

Reply:
I can confirm this is a bug in the expender. Let me work on a fix

Reply:
sent a fix here:
https://github.com/triton-lang/triton/pull/6408

Reply:
Thanks, cherry picked the change and removed the WA here.